src/pkg/retention/res/selectors/label/selector_test.go

```diff
@@ -0,0 +1,121 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package label
+
+import (
+	"github.com/goharbor/harbor/src/pkg/retention/res"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"testing"
+	"time"
+)
+
+// LabelSelectorTestSuite is a suite for testing the label selector
+type LabelSelectorTestSuite struct {
+	suite.Suite
+
+	artifacts []*res.Candidate
+}
+
+// TestLabelSelector is entrance for LabelSelectorTestSuite
+func TestLabelSelector(t *testing.T) {
+	suite.Run(t, new(LabelSelectorTestSuite))
+}
+
+// SetupSuite to do preparation work
+func (suite *LabelSelectorTestSuite) SetupSuite() {
+	suite.artifacts = []*res.Candidate{
+		{
+			NamespaceID:  1,
+			Namespace:    "library",
+			Repository:   "harbor",
+			Tag:          "1.9",
+			Kind:         res.Image,
+			PushedTime:   time.Now().Unix() - 3600,
+			PulledTime:   time.Now().Unix(),
+			CreationTime: time.Now().Unix() - 7200,
+			Labels:       []string{"label1", "label2", "label3"},
+		},
+		{
+			NamespaceID:  1,
+			Namespace:    "library",
+			Repository:   "harbor",
+			Tag:          "dev",
+			Kind:         res.Image,
+			PushedTime:   time.Now().Unix() - 3600,
+			PulledTime:   time.Now().Unix(),
+			CreationTime: time.Now().Unix() - 7200,
+			Labels:       []string{"label1", "label4", "label5"},
+		},
+	}
+}
+
+// TestWithLabelsUnMatched tests the selector of `with` labels but nothing matched
+func (suite *LabelSelectorTestSuite) TestWithLabelsUnMatched() {
+	withNothing := &selector{
+		decoration: With,
+		labels:     []string{"label6"},
+	}
+
+	selected, err := withNothing.Select(suite.artifacts)
+	require.NoError(suite.T(), err)
+	assert.Equal(suite.T(), 0, len(selected))
```

Comment:
We should also probably check that the correct one was selected.

Reply:
Sure